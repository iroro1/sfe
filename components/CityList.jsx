"use client";
import { sortObject } from "@/services/functions";
import { getTopCitiesApi, getWeatherApi } from "@/services/weatherService";
import { Refresh } from "iconsax-react";
import { useRouter } from "next/navigation";
import { useEffect, useState } from "react";
import WeatherCard from "./WeatherCard";
import WeatherCardUser from "./WeatherCardUser";

const CityList = () => {
  const [weather, setWeather] = useState([]);
  const [load, setLoad] = useState(true);
  const [loaded, setLoaded] = useState(false);
  const [userWeather, setUserWeather] = useState({});
  const [location, setLocation] = useState();
  const [topCities, setTopCities] = useState();

  useEffect(() => {
    loadLocation();
  }, []);
  useEffect(() => {
    setTimeout(() => {
      setLoad(false);
    }, 6000);
  }, []);
  useEffect(() => {
    setLoad(true);
    if (typeof window !== undefined) {
      setWeather(JSON.parse(window.localStorage.getItem("citiyData")));
      setUserWeather(JSON.parse(window.localStorage.getItem("userLocation")));
    } else false;
  }, []);
  const loadLocation = async () => {
    setLoad(true);
    await getTopCitiesApi();
    if ("geolocation" in navigator) {
      setLoad(true);
      // Retrieve latitude & longitude coordinates from `navigator.geolocation` Web API
      navigator.geolocation.getCurrentPosition(async ({ coords }) => {
        const { latitude, longitude } = coords;
        const lat = latitude;
        const lng = longitude;
        // setLoad(true);
        try {
          const { data: res } = await getWeatherApi(lat, lng);
          if (res?.location && res?.current) {
            typeof window !== undefined
              ? window.localStorage.setItem(
                  "userLocation",
                  JSON.stringify({
                    location: res?.location,
                    current: res?.current,
                  })
                )
              : false;
          }
        } catch (error) {
          console.log(error);
        }
      });
    }
    if (typeof window !== undefined) {
      const ctData = window.localStorage.getItem("citiyData");
      const uData = window.localStorage.getItem("userLocation");
      if (ctData !== "") setTopCities(JSON.parse(ctData));
      else {
        const { status, data } = getTopCitiesApi();
        if (status === 200) setTopCities(data);
      }
      if (uData !== "") setLocation(JSON.parse(uData));
      else {
      }
    } else false;
    // if ("geolocation" in navigator) {
    //   // Retrieve latitude & longitude coordinates from `navigator.geolocation` Web API
    //   navigator.geolocation.getCurrentPosition(async ({ coords }) => {
    //     const { latitude, longitude } = coords;
    //     const lat = latitude;
    //     const lng = longitude;
    //     try {
    //       const { data: res } = await getWeatherApi(lat, lng);
    //       if (res?.location && res?.current) {
    //         typeof window !== undefined
    //           ? window.localStorage.setItem(
    //               "userLocation",
    //               JSON.stringify({
    //                 location: res?.location,
    //                 current: res?.current,
    //               })
    //             )
    //           : false;
    //         setLocation({
    //           location: res?.location,
    //           current: res?.current,
    //         });
    //       }
    //     } catch (error) {
    //       console.log(error);
    //     }
    //   });
    // }
    setLoad(false);
  };

  const navigate = useRouter();
  const onFav = (name) => {
    const newData = weather.map((item) => {
      if (item.location.name === name) {
        return {
          ...item,
          isFav: !item.isFav,
          name: item?.location?.name,
        };
      } else {
        return { ...item, name: item?.location?.name };
      }
    });
    if (typeof window !== undefined) {
      window.localStorage.setItem("citiyData", JSON.stringify(newData));
    } else false;
    sortObject(newData);
    setWeather(newData);
  };
  const onDelete = (name) => {
    const newData = weather.filter((item) => {
      if (item.location.name !== name) {
        return {
          ...item,
        };
      }
    });
    sortObject(newData);
    setWeather(newData);
  };

  const favoriteList = () => {
    return weather?.filter((item) => item?.isFav);
  };
  const notFavoriteList = () => {
    return weather?.filter((item) => !item?.isFav);
  };
  return load ? (
    <main>
      {" "}
      <h6>Loaing</h6>
    </main>
  ) : (
    <main>
      <div onClick={() => loadLocation()} className="cflex home-reload cmb-18">
        <span>Get Updated data</span>
        <Refresh size={14} className={load && "animate-spin"} />
      </div>
      <div className="main-card">
        {!load && (
          <WeatherCardUser
            load={load}
            data={userWeather}
            onClick={() =>
              navigate.push(
                `/${userWeather?.location?.tz_id}?params=${JSON.stringify(
                  userWeather
                )}`
              )
            }
          />
        )}
      </div>

      <div>
        <h2 className="detail-head">Favorites</h2>
      </div>
      <div className="tab">
        {favoriteList()?.map((city, i) => (
          <div key={i} className="cmb-18 ">
            <WeatherCard
              load={load}
              favClick={() => onFav(city?.location?.name)}
              deleteClick={() => onDelete(city?.location?.name)}
              key={i}
              data={city}
              onClick={() =>
                navigate.push(
                  `/${city?.location?.tz_id}?params=${JSON.stringify(city)}`
                )
              }
            />
          </div>
        ))}
      </div>

      <div>
        <h2 className="detail-head">Cities</h2>
      </div>
      <div className="tab">
        {notFavoriteList()?.map((city, i) => (
          <div key={i} className="cmb-18 ">
            <WeatherCard
              load={load}
              favClick={() => onFav(city?.location?.name)}
              deleteClick={() => onDelete(city?.location?.name)}
              key={i}
              data={city}
              onClick={() =>
                navigate.push(
                  `/${city?.location?.tz_id}?params=${JSON.stringify(city)}`
                )
              }
            />
          </div>
        ))}
      </div>
    </main>
  );
};

export default CityList;
