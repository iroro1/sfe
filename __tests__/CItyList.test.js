/** @jest-environment jsdom */

import "@testing-library/jest-dom";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import React from "react";
import CityList from "../components/CityList";

// Mock the useRouter hook
jest.mock("next/navigation", () => ({
  useRouter: () => ({
    push: jest.fn(),
  }),
}));

// Mock your API functions (getTopCitiesApi and getWeatherApi)
jest.mock("../services/weatherService", () => ({
  getTopCitiesApi: jest.fn(() => Promise.resolve([])),
  getWeatherApi: jest.fn(() => Promise.resolve({})),
}));
describe("CityList Component", () => {
  it("should render loading message when userWeather is undefined", () => {
    const { container } = render(<CityList />);
    expect(screen.getByText("Loading...")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
  it("should render the component with userWeather data", async () => {
    const userWeatherData = {
      location: { name: "Test City", tz_id: "TestTZ" },
      current: { temperature: 25 },
    };
    window.localStorage.setItem(
      "userLocation",
      JSON.stringify(userWeatherData)
    );
    const { container } = render(<CityList />);
    await waitFor(() => {
      expect(screen.getByText("Get Updated data")).toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });
  });
  it('should trigger reload when "Get Updated data" is clicked', async () => {
    const { container } = render(<CityList />);
    const reloadButton = screen.getByText("Get Updated data");
    fireEvent.click(reloadButton);
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });
});
