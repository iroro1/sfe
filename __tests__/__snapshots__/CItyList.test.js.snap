// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CityList Component should render loading message when userWeather is undefined 1`] = `
<div>
  <main>
     
    <h6>
      Loading...
    </h6>
  </main>
</div>
`;

exports[`CityList Component should render the component with userWeather data 1`] = `
<div>
  <main>
    <div
      class="cflex home-reload cmb-18"
    >
      <span>
        Get Updated data
      </span>
      <svg
        class="animate-spin"
        fill="none"
        height="14"
        viewBox="0 0 24 24"
        width="14"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M22 12c0 5.52-4.48 10-10 10s-8.89-5.56-8.89-5.56m0 0h4.52m-4.52 0v5M2 12C2 6.48 6.44 2 12 2c6.67 0 10 5.56 10 5.56m0 0v-5m0 5h-4.44"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.5"
        />
      </svg>
    </div>
    <div
      class="main-card"
    />
    <div>
      <h2
        class="detail-head"
      >
        Favorites
      </h2>
    </div>
    <div
      class="tab"
    />
    <div>
      <h2
        class="detail-head"
      >
        Cities
      </h2>
    </div>
    <div
      class="tab"
    />
  </main>
</div>
`;

exports[`CityList Component should trigger reload when "Get Updated data" is clicked 1`] = `
<div>
  <main>
    <div
      class="cflex home-reload cmb-18"
    >
      <span>
        Get Updated data
      </span>
      <svg
        class="animate-spin"
        fill="none"
        height="14"
        viewBox="0 0 24 24"
        width="14"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M22 12c0 5.52-4.48 10-10 10s-8.89-5.56-8.89-5.56m0 0h4.52m-4.52 0v5M2 12C2 6.48 6.44 2 12 2c6.67 0 10 5.56 10 5.56m0 0v-5m0 5h-4.44"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.5"
        />
      </svg>
    </div>
    <div
      class="main-card"
    />
    <div>
      <h2
        class="detail-head"
      >
        Favorites
      </h2>
    </div>
    <div
      class="tab"
    />
    <div>
      <h2
        class="detail-head"
      >
        Cities
      </h2>
    </div>
    <div
      class="tab"
    />
  </main>
</div>
`;
